### PR TITLE
AzureRMWebAppTask migrated to nodejs

### PR DESCRIPTION
**Pull request by Ajay-MS**
Merging rmmaster2 to master

### CODE REVIEW

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 1]
```diff
@@ -1,8 +1,8 @@
 {
     "id": "497D490F-EEA7-4F2B-AB94-48D9C1ACDCB1",
     "name": "AzureRMWebAppDeployment",
```
Comment (mvvsubbu) [inline Tasks/AzureRmWebAppDeployment/task.json]: Do sit with PM once and get final help markdowns scrubbed

File: Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        if(virtualApplication) {
+            publishingProfile.destinationAppUrl += "/" + virtualApplication;
+        }
+
+        if(webAppUri) {
+            tl.setVariable(webAppUri, publishingProfile.destinationAppUrl);
+        }
+
+        if(canUseWebDeploy(useWebDeploy)) {        
+           tl._writeLine("##vso[task.setvariable variable=websitePassword;issecret=true;]" + publishingProfile.userPWD);
```
Comment (mvvsubbu) [inline Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts]: Mask user id as well

File: Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+    try {
+        tl._writeLine(await azureRmUtil.updateDeploymentStatus(publishingProfile, isDeploymentSuccess));
+    }
+    catch(error) {
+        tl.warning(error);
+    }
+    finally {
+        if(!isDeploymentSuccess) {
+            throw Error(deploymentError);
```
Comment (mvvsubbu) [inline Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts]: Move this code out of finally block as this not at all related to previous try/catch

File: Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
```
Comment (mvvsubbu) [inline Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts]: Move this code out of finally block as this not at all related to previous try/catch

File: Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+ * Checks whether the given path is file or not.
+ * @param path input file path
+ * 
+ * @return true/false based on input is file or not.
+
+ */
+function fileExists(path): boolean {
+  try  {
+    return tl.stats(path).isFile();
+  }
```
Comment (mvvsubbu) [inline Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts]: Post this review can you add this function in vsts-task-lib as well

File: Tasks/AzureRmWebAppDeployment/AzureRMUtil.ts [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+    var webAppHostUrl = webAppPublishKuduUrl.split(':')[0];
+    var requestUrl = "https://" + encodeURIComponent(webAppHostUrl) + "/deployments/" + encodeURIComponent(deploymentId);
+
+    var requestDetails = new Array<string>();
+    requestDetails["requestBody"] = requestBody;
+    requestDetails["requestUrl"] = requestUrl;
+    return requestDetails;
+}
+
+async function getWebAppPublishProfile(SPN, webAppName: string, resourceGroupName: string, deployToSlotFlag: boolean, slotName: string ) {
```
Comment (mvvsubbu) [inline Tasks/AzureRmWebAppDeployment/AzureRMUtil.ts]: This is not required to be separate function merge both functions

File: Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts (hunk 2 above)
Comment (Ajay-MS) [inline Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts]: Done

File: Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts (hunk 3 above)
Comment (Ajay-MS) [inline Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts]: Done

File: Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts (hunk 4 above)
Comment (Ajay-MS) [inline Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts]: Done

File: Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts (hunk 5 above)
Comment (Ajay-MS) [inline Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts]: Will do it.

File: Tasks/AzureRmWebAppDeployment/AzureRMUtil.ts (hunk 6 above)
Comment (Ajay-MS) [inline Tasks/AzureRmWebAppDeployment/AzureRMUtil.ts]: Done

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 7]
```diff
@@ -76,19 +81,47 @@
...
+            "helpMarkDown": "Specify the name of the Virtual Application that has been configured in the Azure portal. The option is not required for deployments to the website root."
+        },
+        {
             "name": "Package",
             "type": "filePath",
-            "label": "Package",
+            "label": "Package or Folder",
             "defaultValue": "",
             "required": true,
             "helpMarkDown": "Location of the Web App zip package on the automation agent or on a UNC path accessible to the automation agent like, \\\\\\\\BudgetIT\\Web\\Deploy\\Fabrikam.zip. Predefined system variables and wild cards like, $(System.DefaultWorkingDirectory)\\\\***.zip` can be also used here."
```
Comment (RoopeshNair) [inline Tasks/AzureRmWebAppDeployment/task.json]: Folder or file path to the Web App package. Variables (Build | Release), wild cards are supported. <br/>
For example, $(System.DefaultWorkingDirectory)\\\\***.zip

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 8]
```diff
@@ -76,19 +81,47 @@
...
         },
         {
+            "name": "WebAppUri",
+            "type": "string",
+            "label": "Web App Hosted Url",
```
Comment (RoopeshNair) [inline Tasks/AzureRmWebAppDeployment/task.json]: Remove Hosted, the text would just be "Web App URL"

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 9]
```diff
@@ -76,19 +81,47 @@
...
+            "required": false,
+            "defaultValue": "",
+            "groupName": "output",
+            "helpMarkDown": "Provide a name for the variable for the webapp hosted url. The variable can be used as $(variableName) to refer to the hosted url of webapp in subsequent tasks like in the PowerShell on Target Machines task."
```
Comment (RoopeshNair) [inline Tasks/AzureRmWebAppDeployment/task.json]: Specify a name for the output variable that is generated for the URL of the Web App. The variable can be consumed in subsequent tasks.

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 10]
```diff
@@ -76,19 +81,47 @@
...
+        },
+        {
+            "name": "UseWebDeploy",
+            "type": "boolean",
+            "label": "Publish using Web Deploy",
+            "required": false,
+            "defaultValue": "true",
+            "groupName": "AdditionalDeploymentOptions",
+            "helpMarkDown": "The Web Deploy options will only work when the automation agent is on Windows and will be ignored on Linux. The options are also ignored for file formats that are not zipped or in a  folders, like Java WAR files. FTP works with automation agent on both Windows and Linux and for zip files, folders, and Java WAR files."
```
Comment (RoopeshNair) [inline Tasks/AzureRmWebAppDeployment/task.json]: "Publish using web deploy options are supported only when using Windows agent. On other platforms, the task relies on Kudu REST APIs (https://github.com/projectkudu/kudu/wiki/REST-API) to deploy the Web App, and following options are not supported"

File: Tasks/AzureRmWebAppDeployment/task.json (hunk 1 above)
Comment (Ajay-MS) [inline Tasks/AzureRmWebAppDeployment/task.json]: Done

**Pull request merged**